File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int answer = 0;
    static int target = 0;
    static int numbers[];
    
    public int solution(int[] numbers, int target) {
        this.target = target;
        this.numbers = numbers;
        
        dfs(0, 0);
        
        return answer;
    }
    
    private static void dfs(int result, int len) {
        if(len == numbers.length) {
            if(result == target) {
                answer++;
            }
            return;
        }
        if(len < numbers.length) {
            dfs(result + numbers[len], len + 1);
            dfs(result - numbers[len], len + 1);
        }    
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    N = len(numbers)
    
    def dfs(idx, result):
        nonlocal answer
        
        if idx == N:
            if target == result:
                answer += 1
            return
        
        dfs(idx + 1, result + numbers[idx])
        dfs(idx + 1, result - numbers[idx])
    
    dfs(0, 0)
    return answer